### ISSUE TITLE
With NOWAIT, values are wrapped in a promise

### ISSUE THREAD

**Issue by canac** [angular-ui/ui-router#3378]
This is a:

- [x] Bug Report
- [ ] Feature Request
- [ ] General Query 

My version of UI-Router is: 1.0.0-rc.1

# Bug Report

#### Current Behavior:

First, let me say that I **love** the flexibility of being able to configure the resolve policy to not wait on promises. However, when using an `async` resolve policy of `NOWAIT`, all values (whether promises or not) are injected as promises.

#### Expected Behavior:

I expect that the value injected into the controller or bound to the component would be the raw value returned from the resolve, i.e. it would not be wrapped in a promise first. If all values from resolve are intended to be wrapped in a promise, I would expect the documentation to explicitly specify this.

#### Link to Plunker that reproduces the issue:

http://plnkr.co/edit/JKDt1SiEOplNoQnareWf

I would expect value to equal `'value'`, not a promise that resolves to `'value'`.

**Comment by christopherthielen** [comment]
Thanks for the plunker.

I guess I don't understand.  How could we inject the unwrapped value into the controller if it's not ready yet?  What if it takes 30 seconds to load?  I guess that's the whole point of NOWAIT: you can use the resolve system for DI, but you work with a promise instead of waiting for the unwrapped value.

All resolves are always processed as promises.  When you return a literal string from the resolve, it's wrapped as a promise, i.e., `Promise.resolve(value)`.  

If you want some resolves to wait and others to nowait, you can specify policy at the resolve level, if you use the [array syntax](https://ui-router.github.io/ng1/docs/latest/interfaces/state.statedeclaration.html#resolve).  Example: http://plnkr.co/edit/H8nuFJUoZ6bx44MX4y8q?p=preview

Closing as `wontfix` because we *should be* injecting a NOWAIT resolve as a promise.

**Comment by canac** [comment]
@christopherthielen Thanks for your response and for your work on this awesome project! I understand what you are saying. What I mean is that I would personally expect a string value to be injected into the controller as a string and a promise to be injected as a promise. It surprises me that *everything* is injected as a promise, whether a string or not. Because that behavior goes against my intuition, I suggest a modification to the documentation to clarify that when using `NOWAIT`, all injected values are wrapped in a promise.

Thanks for the link to the array syntax. I will check into that!

**Comment by christopherthielen** [comment]
> I suggest a modification to the documentation to clarify that when using NOWAIT, all injected values are wrapped in a promise.

Will do, thanks for the feedback!

**Comment by christopherthielen** [comment]
Hmm, it's already documented: https://ui-router.github.io/ng1/docs/latest/interfaces/resolve.resolvepolicy.html#async

**Comment by canac** [comment]
Yes, there is documentation, but that documentation does not clearly explain that when using NOWAIT, all injected values are wrapped in a promise. It talks about how promises are treated, but does not discuss how non-promise values are treated, leaving room for the confusion that I experienced.

**Comment by christopherthielen** [comment]
Ah, understood.
